### PR TITLE
Enhanced simplification rules for Div by a positive constant

### LINKED ISSUE #2308: [ARITH] Enhance CanonicalSimplify((241+48x)/16)
```python
import tvm
x = tvm.var("x")
print(tvm.ir_pass.CanonicalSimplify((240+48*x)/ 16)
print(tvm.ir_pass.CanonicalSimplify((241+48*x)/ 16)
```
produces
```
(15 + (x*3))
((241 + (x*48))/16)
```
where expected result is:
```
(15 + (x*3))
(15 + (x*3))
```

### PR DESCRIPTION
**Pull request by derisavi**
This resolves #2308 
1) In addition to the case in which we can prove the `non_divisible` part is within range `[0, coeff)`, we also handle the case where `non_divisible` part is a constant.
2) Tried to clarify the comments
3) added a unit testcase

### CODE REVIEW

File: tests/python/unittest/test_arith_simplify.py [hunk 1]
```diff
@@ -29,6 +29,9 @@ def test_simplify():
     #  assert tvm.ir_pass.Equal(tvm.ir_pass.CanonicalSimplify(n / (-1)),
     #                           tvm.ir_pass.CanonicalSimplify(-n))
 
+def test_simplify_div():
+    x = tvm.var('x')
+    assert tvm.ir_pass.CanonicalSimplify((241+48*x)/16 - (15 + (x*3))).value == 0
```
Comment (yzhliu) [inline tests/python/unittest/test_arith_simplify.py]: could you also add a non-divisible test case, e.g., ((241 + (x*15))/4)

Comment (derisavi) [reply tests/python/unittest/test_arith_simplify.py]: done.

File: src/arithmetic/canonical.cc [hunk 2]
```diff
@@ -526,6 +558,10 @@ class Canonical::Internal : public IRMutator {
     if (pair.size() == 0) {
       int64_t value = GetConstIntValue(v);
       auto n = make_node<ComExprNode>();
+      // TODO(derisavi) : fix this bug. The following can be done only for Euclidean division/mod.
+      //  Therefore, it's only valid when truncated division/mod is equivalent to Euclidean one,
+      //  that is, if and only if a and v are
+      //  both negative or both positive or a is divisible by v.
```
Comment (derisavi) [inline src/arithmetic/canonical.cc]: @tqchen  (and other folks) can you chime in on this? do you agree?

**Comment by sgrechanik-h** [comment]
Will it simplify `(x*48 + 16)/32  ->  (x*3 + 1)/2`, and if not is it difficult to add?

**Comment by derisavi** [comment]
> Will it simplify `(x*48 + 16)/32 -> (x*3 + 1)/2`, and if not is it difficult to add? Upd: since the result is not a linear equation, it is probably unfixable within this PR

That's right. 

From the compiler point of view, I see one advantage with that simplification: it avoids overflow in some cases. Do you have in mind other uses for that simplification?

**Comment by sgrechanik-h** [comment]
> Do you have in mind other uses for that simplification?

Not really, the only thing that comes to my mind is just the general case when the two expressions come from different sources or from different transformation paths and we benefit from proving them equal. But in practice I haven't (yet) seen any case which would benefit from this particular simplification.

**Pull request closed without merging**